src/blocks/upower.rs

```diff
@@ -0,0 +1,306 @@
+use std::fmt;
+use std::time::{Duration, Instant};
+use std::thread;
+
+use chan::Sender;
+use uuid::Uuid;
+
+use config::Config;
+use errors::*;
+use scheduler::Task;
+use block::{Block, ConfigBlock};
+use widget::{I3BarWidget, State};
+use widgets::text::TextWidget;
+use blocks::dbus::{BusType, Connection, Message, MessageItem, Path};
+use blocks::dbus::arg::Variant;
+
+#[derive(Clone, Copy)]
+enum DeviceType {
+    Unknown,
+    LinePower,
+    Battery,
+    Ups,
+    Monitor,
+    Mouse,
+    Keyboard,
+    Pda,
+    Phone,
+}
+
+impl DeviceType {
+    pub fn to_u32(&self) -> u32 {
+        *self as u32
+    }
+}
+
+impl From<u32> for DeviceType {
+    fn from(id: u32) -> Self {
+        match id {
+            // https://upower.freedesktop.org/docs/Device.html#Device:Type
+            // TODO: derive this automatically.
+            1 => DeviceType::LinePower,
+            2 => DeviceType::Battery,
```

Comment:
If the only device type you care about is this one, why bother will all of this other boilerplate? Just have `Battery` and `Other`.